src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -2761,13 +2767,21 @@ private void EnsureDescriptorsInitialized()
 #if !ES_BUILD_STANDALONE
             Debug.Assert(Monitor.IsEntered(EventListener.EventListenersLock));
 #endif
-            if (m_eventData == null)
-            {
-                // get the metadata via reflection.
-                Debug.Assert(m_rawManifest == null);
-                m_rawManifest = CreateManifestAndDescriptors(this.GetType(), Name, this);
-                Debug.Assert(m_eventData != null);
 
+            if (m_eventData == null && m_rawManifest == null)
+            {
+                if (m_EventMetadataInitializer != null && m_EventDataInitializer != null)
+                {
```

Comment:
The trimmer will not be able to statically figure out that these are not null if source gen is available for this event source. The result will be that it will still keep the reflection based code in `CreateManifestAndDescriptors`.

I think the best way to avoid this problem would be to introduce a new protected .ctor, which will be designed for the source generators (no need for protected properties/fields anymore). And then internally we'll refactor the event source to not have a reference to `CreateManifestAndDescriptors` from anywhere but the non-source-generators .ctors. That way if all event sources in the app use the source generators, trimmer will not see a reference to `CreateManifectAndDescriptors` and will be able to remove it.

Reply:
We're adding internal fields which act like the new public API in a way. I'm fine if we make the new `.ctor` internal for now. I'm also perfectly OK if we decide to not do the internal refactoring needed for trimming to work. But I think it's definitely worthwhile to use the correct API pattern now (even if it's only internal).

If perf is the driving goal here, it's likely we will make the new fields protected soon, so that we can use them from non-CoreLib assemblies... and if we really want to fix perf, we would have to expose them via ref assemblies so that ASP.NET can use them as well. Now they're true public API - and we missed the boat on doing this in a way which would allow us to trim the internals in the future. I would very much like to avoid that happening.

The change is pretty small - instead of using effectively protected fields (which I honestly find to be a somewhat weird), use a `.ctor` designed for this scenario. That's it. The rest of the changes necessary for trimming will then all be made internally and can be easily postponed.

Reply:
> instead of using effectively protected fields (which I honestly find to be a somewhat weird), use a `.ctor` designed for this scenario. 

A virtual protected property will allow use of `ReadOnlySpan<byte>` which can then be a pointer to data section; which couldn't done via a `.ctor` alone

Reply:
I must admit I don't know why - property is effectively a method (setter), so how is it different from `.ctor` (which is also a method)?

But if we need some of the data to be passed via a property, that's OK - we just need to have some pattern where:
* Source generator calls NoReflectionEntryPoint
* All other ways (existing API) calls ReflectionEntryPoint

That way we can "hide" all of the reflection into the ReflectionEntryPoint and trimmer will be able to remove it, if only source generators are used.
Patterns which are based on setting values to fields or parameters are currently not supported by the trimmer (and are very hard to do, so I would not rely on them being supported any time soon, if ever).

Reply:
> I must admit I don't know why - property is effectively a method (setter), so how is it different from .ctor (which is also a method)?

Its only the getter that can work this way; so it would be a readonly Property with only `{ get; }` so nothing would be passed to it; but generated as C#, much like `ProviderMetadata` works with the current generator:

```csharp
partial class RuntimeEventSource
{
    private RuntimeEventSource() : base(new Guid(0x49592c0f,0x5a05,0x516d,0xaa,0x4b,0xa6,0x4e,0x02,0x02,0x6c,0x89), "System.Runtime") { }

    private protected override ReadOnlySpan<byte> ProviderMetadata => new byte[] { 0x11, 0x0, 0x53, 0x79, 0x73, 0x74, 0x65, 0x6d, 0x2e, 0x52, 0x75, 0x6e, 0x74, 0x69, 0x6d, 0x65, 0x0, };
}
```

Reply:
Don't know if the official docs are there; but @vcsjones's blog on the `ReadOnlySpan<byte>` compiler optimization is quite good  https://vcsjones.dev/csharp-readonly-span-bytes-static/

Reply:
> From a work scoping perspective I think we should separate out linkability from CoreCLR startup performance improvement. ... I suggest we draw the line at not doing any changes (for the moment) that touch public API surface area.

I agree.  I suggest we keep this scoped to performance improvement in .NET 6 and then work towards linkability in .NET 7.

Reply:
Yes, I'm trying to minimize any public API changes for now unless it is absolutely necessary.

> But I think it's definitely worthwhile to use the correct API pattern now (even if it's only internal).

I agree. I'll look into adding the suggested pattern here via NoReflectionEntryPoint and ReflectionEntryPoint. Is there any examples that do this in BCL?

Reply:
> Now they're true public API - and we missed the boat on doing this in a way which would allow us to trim the internals in the future. I would very much like to avoid that happening.

Fully agree. I don't want to make the public API changes until we are confident we won't regret them. If that means a chunk (perhaps event a large chunk) of the perf improvements will be out of reach for the moment I think that is an OK outcome.

Reply:
> Yes, I'm trying to minimize any public API changes for now unless it is absolutely necessary

I'd be tempted to go further, if public API changes appear to be needed that should be a warning sign that the scope of work has grown too large and we need to scale it back. I'd rather err on the side of having less perf improvement right now and keeping the work scoped.